solutions/python-igraph/clique_percolation/clique_percolation.py

```diff
@@ -0,0 +1,88 @@
+import igraph as ig
+import matplotlib.pyplot as plt
+
+EXAMPLE_K = 3
+
+def k_cliques(g, k):
+    """Returns a list of all cliques of size k in Graph g. Each clique is given as a set of vertex ids in g.
+    """
+    return g.cliques(k, k)
+
+def k_communities(g, k, min_common_vertices):
+    """
+    Given a graph and size k, return a list of communities found through clique
+    percolation with at least min_common_vertices shared.
+    A community is the connected component of cliques where two cliques overlap
+    in at least k-1 vertices.
+    Each community is a set of vertices in that community
+
+    Approach: generate a new clique graph where each k-clique is a vertex, and 
+    connect vertices if the cliques share min_common_vertices
+    """
+    if (min_common_vertices > k):
+        raise ValueError("Error: min_common_vertices=%d must be greater than k=%d" % (min_common_vertices, k))
+
+    # Each clique can be identified by it's index in 'cliques'
+    cliques = [set(x) for x in k_cliques(g, k)]
+    num_cliques = len(cliques)
+    # Graph storing each clique as a vertex
+    clique_graph = ig.Graph()
+    clique_graph.add_vertices(num_cliques)
+    cg_edges = []
+    # List of communities as a list of list of clique indices
+    clique_communities = []
+    # List of communities as a list of list of set of vertices
+    communities = []
+
+    # Get connected cliques
+    for c1 in range(num_cliques):
+        for c2 in range(c1, num_cliques):
+            # If the two cliques share at least min_common_vertices,
+            # they are part of the same community
+            if c1 != c2 and len(cliques[c1].intersection(cliques[c2])) >= min_common_vertices:
+                cg_edges.append((c1, c2))
```

Comment:
This double for loop can be simplified to a one-liner using `itertools.combinations()` and Python list comprehensions.